cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -654,4 +671,53 @@ public static void setupDatasets(DatasetFramework framework) throws IOException,
     framework.addInstance(MetadataDataset.class.getName(), BUSINESS_METADATA_INSTANCE_ID, DatasetProperties.EMPTY);
     framework.addInstance(MetadataDataset.class.getName(), SYSTEM_METADATA_INSTANCE_ID, DatasetProperties.EMPTY);
   }
+
+  @Override
+  public void markUpgradeState(MetadataScope scope) throws DatasetManagementException, IOException {
+    DatasetId datasetId = getMetadataDatasetInstance(scope);
+    LOG.info("Add Upgrade tag with version {} to {}", ProjectInfo.getVersion().toString(), datasetId);
+    addTags(scope, datasetId, getTagWithVersion(ProjectInfo.getVersion().toString()));
+    removeTags(scope, datasetId, NEEDS_UPGRADE_TAG);
+  }
+
+  @Override
+  public boolean isUpgradeRequired(MetadataScope scope) throws DatasetManagementException {
+    DatasetId datasetId = getMetadataDatasetInstance(scope);
+    if (!dsFramework.hasInstance(datasetId)) {
+      return true;
```

Comment:
really? It does not exist, so why upgrade? In this case, it only needs to be created.